### ISSUE TITLE
Complex bug in `__init__` overload in generic class, in pyright 1.1.309

### ISSUE THREAD

**Issue by math-fehr** [microsoft/pyright#5144]
**Describe the bug**
It is quite hard to find the exact cause of the bug and changing anything in my code
seems to remove it. It seems that when overriding an `__init__` on a generic class,
that inherit from a dataclass, the wrong overload is chosen when reassigning to a
variable. There are other conditions that seems to make this fail, I noted them in the
example.

**To Reproduce**
Use pyright 1.1.309, with the command line.

**Expected behavior**
No errors (which is the case in pyright 1.1.308).

**Screenshots or Code**
Here is a reduced version of our problem. I couldn't manage to reduce it further,
any change reducing the code seemed to remove that bug.

```python
from __future__ import annotations
from dataclasses import dataclass
from typing import Generic, TypeVar, overload


# If `Parent` is not a dataclass, or if the dataclass is set to `init=False`,
# the bug goes away, even if `Parent` defines its own `__init__`.
@dataclass
class Parent:
    pass


# If T is replaced by `float`, the bug goes away.
class T:
    pass


_FooTyp = TypeVar("_FooTyp", bound=T | str, covariant=True)


class Foo(Generic[_FooTyp], Parent):
    # Overload of __init__
    @overload
    def __init__(self: Foo[_FooTyp], arg: _FooTyp) -> None:
        ...

    @overload
    def __init__(self: Foo[str], arg: int) -> None:
        ...

    def __init__(self, arg: int | T | str) -> None:
        pass


def foo(value: int | Foo[T | str]):
    if isinstance(value, int):
        # Interestingly, the error is not triggered if we assign this to another variable.
        value = Foo(32)  # error:
        # Argument of type "Literal[32]" cannot be assigned to parameter "arg" of type "T | str" in function "__init__"
        #    Type "Literal[32]" cannot be assigned to type "T | str"
        #    "Literal[32]" is incompatible with "T"
        #    "Literal[32]" is incompatible with "str"
 ```

**Comment by erictraut** [comment]
Type validation of a constructor is complicated because it potentially involves multiple functions (the metaclass's `__call__` method, the `__new__` method and the `__init__` method) and all of these methods affect each other (e.g. the output type of the `__new__` method affects which `__init__` method is used).

Until 1.1.309, pyright's constructor validation did things kind of backward. It started by validating the `__init__` method for the class, then it validated the `__new__` method if the class hierarchy provided one. This is problematic in cases where the `__new__` method constructs an instance of a different class. (Yes, that case is rare, but it does happen in some Python code.) The new constructor validation logic introduced in 1.1.309 performs type validation in the proper order, more accurately reflecting the runtime behavior.

Let's simplify your code slightly to explore what's happening.

```python
from __future__ import annotations
from typing import Generic, Self, TypeVar, overload

class Bar: ...

_FooTyp = TypeVar("_FooTyp", bound=Bar | str, covariant=True)

class Foo(Generic[_FooTyp]):
    def __new__(cls, *args, **kwargs) -> Self:
        return super().__new__(cls, *args, **kwargs)

    @overload
    def __init__(self: Foo[_FooTyp], arg: _FooTyp) -> None: ...

    @overload
    def __init__(self: Foo[str], arg: int) -> None: ...

    def __init__(self, arg: int | Bar | str) -> None:
        pass

value: Foo[Bar | str] = Foo(32)
```

I removed the `Parent` class and the `foo` function. I also renamed `T` to `Bar` because my brain thinks of `T` as a `TypeVar`. I also introduced a custom `__new__` method that overrides the standard `object.__new__`. This is the same `__new__` method signature that's synthesized by a dataclass, which explains why you were seeing the problem only if you included a `@dataclass` decorator.

Let's focus on the last line of code in the sample: `value: Foo[Bar | str] = Foo(32)`. This will use bidirectional type inference to determine the return type of the constructor call `Foo(32)`. Imagine for a moment that there's no `__init__` method, and only the `__new__` method is present in `Foo`. I think we'd all agree that the correct inferred return result of the `__new__` method in this case would be `Foo[Bar | str]`. 

Next, let's consider the `__init__` method. If we assume that `__new__` constructs an object of type `Foo[Bar | str]`, then the `__init__` method is specialized as follows:

```python
    @overload
    def __init__(self: Foo[Bar | str], arg: Bar | str) -> None: ...

    @overload
    def __init__(self: Foo[str], arg: int) -> None: ...
```

Since we're assuming that the `__new__` method constructs a `Foo[Bar | str]`, only the first of the two overloads is compatible based on the `self` parameter type. The argument `32` supplied for the `arg` parameter isn't compatible with `Bar | str` in the first overload, hence the type violation.

I'll need to give this issue more thought. I haven't convinced myself that the above behavior is correct — and if not, how it can be adjusted without breaking other legitimate cases.

For comparison, mypy accepts this code, but mypy also has many known issues with constructor validation. For example, it doesn't attempt to handle the case where `__new__` returns an instance of another class. I don't think I can use mypy as a guide here to inform the "correct" behavior.

Let me know if you have any thoughts on my analysis.

**Comment by erictraut** [comment]
Incidentally, if you want a quick workaround, you can change the constructor call to `Foo[str](32)`. By explicitly specializing `Foo` with a type argument of `str`, you eliminate the issue.

**Comment by math-fehr** [comment]
As always, thanks a lot for your fast and detailed explanation!
This makes a lot of sense, and I guess the problem was that I was using `__init__` itself to infer the type of the created object, which is not the correct way anymore.
I think that another solution would be to overload `__new__` as well, which would give me the correct behavior.

Taking your example, it would look like this, which is now removing the error:
```python
from __future__ import annotations
from typing import Generic, TypeVar, overload


class Bar:
    ...


_FooTyp = TypeVar("_FooTyp", bound=Bar | str, covariant=True)


class Foo(Generic[_FooTyp]):
    @overload
    def __new__(cls, arg: _FooTyp) -> Foo[_FooTyp]:
        ...

    @overload
    def __new__(cls, arg: int) -> Foo[str]:
        ...

    def __new__(cls, *args, **kwargs) -> Foo[str] | Foo[Bar | str]:
        return super().__new__(cls, *args, **kwargs)

    @overload
    def __init__(self: Foo[_FooTyp], arg: _FooTyp) -> None:
        ...

    @overload
    def __init__(self: Foo[str], arg: int) -> None:
        ...

    def __init__(self, arg: int | Bar | str) -> None:
        pass


value: Foo[Bar | str] = Foo(32)
```

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.311, which I just published. It will also be included in a future release of pylance.
